src/tokens.c

```diff
@@ -0,0 +1,219 @@
+#include "tokens.h"
+#include "bufView.h"
+#include "textUtils.h"
+#include "hash.h"
+#include "bech32.h"
+
+#define ASSET_FINGERPRINT_SIZE 20
+
+void deriveAssetFingerprintBytes(
+        const uint8_t* policyId,
+        size_t policyIdSize,
+        const uint8_t* assetName,
+        size_t assetNameSize,
+        uint8_t *fingerprintBuffer,
+        size_t fingerprintBufferSize
+)
+{
+	ASSERT(policyIdSize == MINTING_POLICY_ID_SIZE);
+	ASSERT(assetNameSize <= ASSET_NAME_SIZE_MAX);
+	ASSERT(fingerprintBufferSize >= ASSET_FINGERPRINT_SIZE);
+
+	uint8_t hashInput[MINTING_POLICY_ID_SIZE + ASSET_NAME_SIZE_MAX] = {0};
+	const size_t hashInputSize = policyIdSize + assetNameSize;
+	{
+		write_view_t view = make_write_view(hashInput, hashInput + SIZEOF(hashInput));
+		view_appendBuffer(&view, policyId, policyIdSize);
+		view_appendBuffer(&view, assetName, assetNameSize);
+		ASSERT(view_processedSize(&view) == hashInputSize);
+	}
+
+	blake2b_160_hash(hashInput, hashInputSize, fingerprintBuffer, fingerprintBufferSize);
+}
+
+size_t deriveAssetFingerprintBech32(
+        const uint8_t* policyId,
+        size_t policyIdSize,
+        const uint8_t* assetName,
+        size_t assetNameSize,
+        char* fingerprint,
+        size_t fingerprintMaxSize
+)
+{
+	ASSERT(policyIdSize == MINTING_POLICY_ID_SIZE);
+	ASSERT(assetNameSize <= ASSET_NAME_SIZE_MAX);
+
+	uint8_t fingerprintBuffer[ASSET_FINGERPRINT_SIZE];
+	deriveAssetFingerprintBytes(
+	        policyId, policyIdSize,
+	        assetName, assetNameSize,
+	        fingerprintBuffer, SIZEOF(fingerprintBuffer)
+	);
+
+	size_t len = bech32_encode("asset", fingerprintBuffer, SIZEOF(fingerprintBuffer), fingerprint, fingerprintMaxSize);
+	ASSERT(len == strlen(fingerprint));
+	ASSERT(len + 1 < fingerprintMaxSize);
+
+	return len;
+}
+
+typedef struct {
+	uint8_t fingerprint[ASSET_FINGERPRINT_SIZE];
+	uint8_t places;
+} token_info_t;
+
+const token_info_t tokenInfos[] = {
```

Comment:
Would it be useful to also generate comments describing the underlying policy id and asset name for each fingerprint?